### PR TITLE
167 allow renaming

### LINKED ISSUE #167: Allow renaming of entities
It's quite common to think of a better name for a model / protocol further down the line

### PR DESCRIPTION
**Pull request by sroderick-g5sro**
Will now fix #167.

### CODE REVIEW

File: weblab/entities/forms.py
Comment (sroderick-g5sro) [inline weblab/entities/forms.py:111]: Ideally would say if it was a model or protocol but unsure how to get the display type here

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:793]: `is_editable_by` would probably be better than `is_managed_by` here, assuming we want to allow collaborators with write access to rename entities @mirams ?

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:785]: Similarly to `EntityCreateView`, you can have 2 subclasses of `EntityRenameForm`, one for models and one for protocols (and a third for `FittingSpec` in the `fitting` app actually). And rather than defining `form_class` like this, make it an `@property` which returns the form class corresponding to the class `self.model`.

File: weblab/entities/views.py
Comment (jonc125) [reply weblab/entities/views.py:793]: Which means you could just inherit from `EntityCollaboratorRequiredMixin` instead of `UserPassesTestMixin` and remove this method.

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:787]: The `DetailView` class sets `self.object` so we don't need this method.

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:804]:
```suggestion
            entity = self.object
```

File: weblab/entities/forms.py
Comment (jonc125) [reply weblab/entities/forms.py:116]: Indeed, it's rather awkward! The trick is used by `EntityCreateView` (see also below) and the two subclasses of `EntityForm` - you basically need a form subclass for each entity type.

**Review by jonc125 — CHANGES_REQUESTED** [review]
I can confirm that this works nicely on my system.

I've added a few comments, particularly around how to make the name uniqueness check specific to the entity type. Since it's fine to own a model & protocol with the same name, just not 2 models or 2 protocols.

The other thing needed before this can be merged is a few automated tests: for each of a model and a protocol, check both successful rename and error message case.

Then probably merge this PR, and do renaming of fitting specs & datasets in a separate PR to close the issue? Or those features can be added to this PR if you prefer.

File: weblab/entities/views.py
Comment (sroderick-g5sro) [reply weblab/entities/views.py:787]: When I try and take this out I get an error message saying the method is required

Comment (jonc125) [reply weblab/entities/views.py:787]: That might be just because it's used in the `post` and `test_func` methods below? In which case implementing the other changes would remove the need for it. Or it might be that one of the base classes needs it, in which case leave it in.

File: weblab/entities/forms.py
Comment (jonc125) [inline weblab/entities/forms.py:10]: You can get rid of this import now. There are a few other flake8 errors in the build too; I'll not go through them all.

File: weblab/entities/forms.py
Comment (jonc125) [inline weblab/entities/forms.py:44]: I thought we might need a `, author=self.user` argument in the filter too, but `EntityForm` doesn't have this. So I think there may be a missing test case for creation too, that checks it's OK for 2 users to have a model with the same name!

File: weblab/entities/tests/test_views.py
Comment (jonc125) [inline weblab/entities/tests/test_views.py:107]: Yeah, following from the comment above: this shouldn't fail, because `model2` has a different author from `model`.

File: weblab/entities/tests/test_views.py
Comment (jonc125) [inline weblab/entities/tests/test_views.py:148]: `model` here won't actually reflect changes in the database. So instead you want to check `entity.repo_abs_path` after checking the name.

File: weblab/entities/tests/test_views.py
Comment (jonc125) [inline weblab/entities/tests/test_views.py:155]: Oops!

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:983]:
```suggestion
                'bives request failed: %d (%s)' %
                (bives_response.status_code, bives_response.content.decode())
```

File: weblab/entities/forms.py
Comment (sroderick-g5sro) [reply weblab/entities/forms.py:44]: sounds like we do need to filter by username in the form so I have added `author=self.user` as a filter option - is this correct?

Comment (sroderick-g5sro) [reply weblab/entities/forms.py:44]: interestingly I now get a database integrity error thrown, when I try to a rename model to one that exists. Not sure if I need to trap this, was just expecting the message from the form
 
image

File: weblab/entities/tests/test_views.py
Comment (sroderick-g5sro) [reply weblab/entities/tests/test_views.py:148]: thanks, change made

File: weblab/entities/forms.py
Comment (jonc125) [reply weblab/entities/forms.py:44]: Interesting. `Entity` definitely has `unique_together = ('entity_type', 'name', 'author')` so should allow same name different author. And it looks like from the error it's complaining that the author also matches. So as you say I'd expect the form validation to catch this before it hits the DB. Could you push/paste your work-in-progress so I can take a look?

Comment (jonc125) [reply weblab/entities/forms.py:44]: The test looks OK, so it might be something in this file that's wrong, but I don't think you've pushed any changes here?

File: weblab/entities/forms.py
Comment (jonc125) [inline weblab/entities/forms.py:45]: This looks like what I'd expect... so not sure why it isn't working! Maybe put some print calls in to see whether it's actually being used, and what the DB contains here if it is?

Comment (sroderick-g5sro) [reply weblab/entities/forms.py:45]: thanks Jonathan, it appears to run this line fine.

Comment (sroderick-g5sro) [reply weblab/entities/forms.py:45]: in the database are the two models my model1, test model 2 so still not clear why getting the database constraint issue

Comment (jonc125) [reply weblab/entities/forms.py:45]: It looks like `self.user` is `None` in the form, so the filter always returns false. The patch below resolves the issue; I can't commit it locally because of flake8 issues elsewhere! Basically you need to add `UserFormKwargsMixin` as a superclass for the view so it passes the user through to the form.

```diff
diff --git c/weblab/entities/tests/test_views.py i/weblab/entities/tests/test_views.py
index b1d0c81..31042b4 100644
--- c/weblab/entities/tests/test_views.py
+++ i/weblab/entities/tests/test_views.py
@@ -146,9 +146,9 @@ class TestEntityRenaming:
             data={
                      'name': 'test model 2'
                  })
-        assert response.status_code == 302
+        assert response.status_code == 200
         entity = ModelEntity.objects.first()
-        assert entity.name == 'test model 2'
+        assert entity.name == 'my model1'

     def test_model_and_protocol_renaming_success(self, client, logged_in_user, helpers):
         helpers.add_permission(logged_in_user, 'create_model')
@@ -186,7 +186,7 @@ class TestEntityRenaming:
 @pytest.mark.django_db
 @pytest.mark.parametrize("recipe,url,list_url", [
     (recipes.model, '/entities/models/%d/delete', '/entities/models/'),
-    (recipes.protocol, '/entities/protocols/%d/delete', '/entit1ies/protocols/'),
+    (recipes.protocol, '/entities/protocols/%d/delete', '/entities/protocols/'),
 ])
 class TestEntityDeletion:
     def test_owner_can_delete_entity(
diff --git c/weblab/entities/views.py i/weblab/entities/views.py
index d93fa89..c87005c 100644
--- c/weblab/entities/views.py
+++ i/weblab/entities/views.py
@@ -779,7 +779,7 @@ class EntityFileDownloadView(EntityTypeMixin, EntityVersionMixin, SingleObjectMi
         return response


-class RenameView(LoginRequiredMixin, UserPassesTestMixin, FormMixin, EntityTypeMixin, DetailView):
+class RenameView(LoginRequiredMixin, UserPassesTestMixin, EntityTypeMixin, UserFormKwargsMixin, FormMixin, DetailView):
     template_name = 'entities/entity_rename_form.html'
     context_object_name = 'entity'

```

File: weblab/datasets/forms.py
Comment (jonc125) [inline weblab/datasets/forms.py:35]: There's no need to have subclasses here; this can just be `DatasetRenameForm`

File: weblab/datasets/views.py
Comment (jonc125) [inline weblab/datasets/views.py:254]: `is_managed_by` might be better, although it's the same effect! `is_editable_by` doesn't yet work for datasets due to a quirk in Django's permission system, because they're not in the entities app. See also #247.

File: weblab/fitting/forms.py
Comment (jonc125) [inline weblab/fitting/forms.py:30]: You don't want to repeat this class here, but import it from `entities.forms`.

File: weblab/fitting/urls.py
Comment (jonc125) [inline weblab/fitting/urls.py:137]: I think this should be:
```suggestion
        entity_views.RenameView.as_view(),
```

Comment (jonc125) [reply weblab/fitting/urls.py:137]: Actually no, because you need a different form. So you'll need to create a `FittingSpecRenameView` class in `fitting/views.py` that inherits from `entities.views.RenameView`, and reference `views.FittingSpecRenameView` here.

File: weblab/datasets/tests/test_views.py
Comment (jonc125) [inline weblab/datasets/tests/test_views.py:117]:
```suggestion
class TestDatasetRenaming:
```

File: weblab/datasets/views.py
Comment (jonc125) [inline weblab/datasets/views.py:262]:
```suggestion
            dataset = self._get_object()
```

File: weblab/datasets/views.py
Comment (jonc125) [inline weblab/datasets/views.py:271]: Probably better for this to redirect to the renamed dataset. So
```suggestion
        return reverse(ns + ':detail', args=[self._get_object().id])
```
I think.

File: weblab/entities/tests/test_views.py
Comment (jonc125) [inline weblab/entities/tests/test_views.py:54]: I don't think this is checking anything new at present, as it doesn't create two models? We probably want to defer the fixing of the bug on entity creation for another PR, so I'd suggest removing this test for now.

File: weblab/datasets/tests/test_views.py
Comment (jonc125) [inline weblab/datasets/tests/test_views.py:131]: We should check that after the rename the dataset's archive still can be found on disk, since it behaves differently from models & protocols. So here we want
```python
        assert dataset.archive_path.exists()
```
Which means you need a dataset with a file in the first place. So rather than creating a dataset in the test I suggest using the `my_dataset_with_file` fixture (and so no need for `logged_in_user` or `helpers`). And assert that `dataset.archive_path.exists()` before the rename too just in case!

File: weblab/datasets/views.py
Comment (jonc125) [inline weblab/datasets/views.py:263]: You'll probably need to implement the archive name here. Something like:
```suggestion
            old_archive_path = dataset.archive_path
            dataset.name = new_name
            if old_archive_path.exists():
                old_archive_path.rename(dataset.archive_path)
```

File: weblab/entities/views.py
Comment (jonc125) [inline weblab/entities/views.py:803]:
```suggestion
        """Check the form and possibly rename the entity.
```

**Review by jonc125 — CHANGES_REQUESTED** [review]
Looking good! There are a couple of things that could do with checking before merging though.

File: weblab/datasets/tests/test_views.py
Comment (sroderick-g5sro) [reply weblab/datasets/tests/test_views.py:134]: done - thanks

File: weblab/datasets/tests/test_views.py
Comment (jonc125) [inline weblab/datasets/tests/test_views.py:121]: We could make this test slightly stronger with this change and the next below.
```suggestion
        old_path = dataset.archive_path
        assert old_path.exists()
```

File: weblab/datasets/tests/test_views.py
Comment (jonc125) [inline weblab/datasets/tests/test_views.py:132]:
```suggestion
        assert not old_path.exists()
        assert dataset.archive_path.exists()
```

**Review by jonc125 — CHANGES_REQUESTED** [review]
This looks good in terms of functionality. We should just add tests for the fitting spec case, in `fitting/tests/test_views.py`, along the same lines as those added to `entities/tests/test_views.py`.

**Comment by sroderick-g5sro** [comment]
> This looks good in terms of functionality. We should just add tests for the fitting spec case, in `fitting/tests/test_views.py`, along the same lines as those added to `entities/tests/test_views.py`.

File: weblab/fitting/tests/test_views.py
Comment (jonc125) [inline weblab/fitting/tests/test_views.py:3]:
```suggestion
from core import recipes
```
is used elsewhere

**Review by jonc125 — APPROVED** [review]
Looks ready to merge!

**Pull request merged**
